R/randomForest_tidiers.R

```diff
@@ -684,16 +684,188 @@ rf_partial_dependence <- function(df, ...) { # TODO: write test for this.
   res
 }
 
+ubSMOTE2 <- function(X,Y, max_synth_perc=200, target_minority_perc=40, target_size=NULL, k=5, ...) {
+  if(!is.factor(Y)) 
+    stop("Y has to be a factor")
+  if(is.vector(X)) 
+    stop("X cannot be a vector")  
+  
+  data<-cbind(X,Y)
+  id.1 <- which(Y == 1)
+
+  # Get minority size
+  minority_size = length(id.1)
+  # Get majoirty size
+  majority_size = NROW(data) - minority_size
+
+  row.has.na<-function(X)
+    return(apply(X,1,function(x){any(is.na(x))}))
+
+  smote_minority <- function(data, synth_perc, k) {
+    if (synth_perc < 100) { # ubSmoteExs throws error if synth_perc is less than 100.
+                            # Work it around by synthesizing at least 100% and sample then.
+      synth_perc_ <- 100
+    }
+    else {
+      synth_perc_ <- synth_perc
+    }
+    newExs <- unbalanced::ubSmoteExs(data[id.1,], "Y", synth_perc_, k)   
+    row.is.na<-row.has.na(newExs)
+    
+    if(any(row.is.na)) {
+      newExs<-newExs[!row.is.na, ]
+      colnames(newExs)<-colnames(data)
+      cat("WARNING: NAs generated by SMOTE removed \n")
+    }
+
+    if (synth_perc < 100) { # ubSmoteExs throws error if synth_perc is less than 100.
+                            # Work it around by sampling.
+      nrow_to_synth <- as.integer(nrow(data) * synth_perc / 100)
+      if (nrow(newExs) > nrow_to_synth) {
+        newExs <- newExs %>% sample_rows(nrow_to_synth)
+      }
+    }
+    newExs <- newExs %>% dplyr::mutate(synthesized = TRUE)
+    newExs
+  }
+
+  sample_majority <- function(data, size) {
+    # get the undersample of the "majority class" examples
+    selMaj <- sample((1:NROW(data))[-id.1], size, replace=F)
+
+    ret <- data[selMaj,]
+    ret
+  }
+
+  sample_minority <- function(data, size) {
+    # get the undersample of the "minority class" examples
+    selMin <- sample((1:NROW(data))[id.1], size, replace=F)
+
+    ret <- data[selMin,]
+    ret
+  }
+
+  if (is.null(target_size)) {
+    if (minority_size / (minority_size + majority_size) >= target_minority_perc / 100) {
+      # Already enough minority for the ratio even without SMOTE.
+      # No Action. Already above target minority ratio.
+      newdataset <- data %>% dplyr::mutate(synthesized = FALSE)
+    }
+    else if (minority_size * (100 + max_synth_perc) / 100 / (minority_size * (100 + max_synth_perc) / 100 + majority_size) >= target_minority_perc / 100) {
```

Comment:
this if /else condition is little tricky since it does a lot of calculation in it.  Some comment explaining what's actually is calculated in the condition help to understand it. Same for all the other similar if/else conditions.